File: go/adxl355/device.go
package adxl355

// Device is the ADXL355 accelerometer driver.
type Device struct {
	transport   Transport
	rangeMode   Range
	initialized bool
}

// New creates a new ADXL355 device instance.
func New(transport Transport) *Device {
	return &Device{
		transport:   transport,
		rangeMode:   Range4G,
		initialized: false,
	}
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

func (d *Device) readU8(reg byte) (byte, error) {
	data, err := d.transport.ReadRegister(reg, 1)
	if err != nil {
		return 0, err
	}
	return data[0], nil
}

func (d *Device) writeU8(reg byte, val byte) error {
	return d.transport.WriteRegister(reg, []byte{val})
}

// ---------------------------------------------------------------------------
// Core API
// ---------------------------------------------------------------------------

// Probe verifies device identity by reading ID registers.
func (d *Device) Probe() (bool, error) {
	idAd, err := d.readU8(RegDEVID_AD)
	if err != nil {
		return false, err
	}
	idMst, err := d.readU8(RegDEVID_MST)
	if err != nil {
		return false, err
	}
	partId, err := d.readU8(RegPARTID)
	if err != nil {
		return false, err
	}

	if idAd != DEVID_AD_VALUE || idMst != DEVID_MST_VALUE || partId != PARTID_VALUE {
		return false, ErrBadDevice
	}

	// Enter standby mode after probe
	if err := d.writeU8(RegPOWER_CTL, byte(PowerStandby)); err != nil {
		return false, err
	}

	d.initialized = true
	return true, nil
}

// Reset performs a software reset.
func (d *Device) Reset() error {
	if err := d.writeU8(RegRESET, RESET_CODE); err != nil {
		return err
	}
	d.transport.DelayMs(10)
	d.rangeMode = Range4G
	return nil
}

// SetRange sets the acceleration range.
func (d *Device) SetRange(r Range) error {
	if r < Range2G || r > Range8G {
		return ErrInvalidArg
	}
	if err := d.writeU8(RegRANGE, byte(r)); err != nil {
		return err
	}
	d.rangeMode = r
	return nil
}

// GetRange reads the currently configured range from hardware.
func (d *Device) GetRange() (Range, error) {
	val, err := d.readU8(RegRANGE)
	if err != nil {
		return 0, err
	}
	return Range(val & 0x03), nil
}

// SetPowerMode sets the power mode (standby/measurement).
func (d *Device) SetPowerMode(mode PowerMode) error {
	reg, err := d.readU8(RegPOWER_CTL)
	if err != nil {
		return err
	}
	if mode == PowerMeasurement {
		reg |= 1
	} else {
		reg &^= 1
	}
	return d.writeU8(RegPOWER_CTL, reg)
}

// ReadRaw reads raw 20-bit acceleration data for all three axes.
func (d *Device) ReadRaw() (*RawXYZ, error) {
	data, err := d.transport.ReadRegister(RegXDATA3, 9)
	if err != nil {
		return nil, err
	}
	return &RawXYZ{
		X: DecodeRaw20(data[0], data[1], data[2]),
		Y: DecodeRaw20(data[3], data[4], data[5]),
		Z: DecodeRaw20(data[6], data[7], data[8]),
	}, nil
}

// ReadG reads acceleration in g.
func (d *Device) ReadG() (*AccelXYZ, error) {
	raw, err := d.ReadRaw()
	if err != nil {
		return nil, err
	}
	scale := rangeToScale(d.rangeMode)
	return &AccelXYZ{
		X: float32(raw.X) * scale,
		Y: float32(raw.Y) * scale,
		Z: float32(raw.Z) * scale,
	}, nil
}

// ReadMps2 reads acceleration in m/s².
func (d *Device) ReadMps2() (*AccelXYZ, error) {
	accel, err := d.ReadG()
	if err != nil {
		return nil, err
	}
	return &AccelXYZ{
		X: accel.X * StandardGravityMS2,
		Y: accel.Y * StandardGravityMS2,
		Z: accel.Z * StandardGravityMS2,
	}, nil
}

// ReadTemperatureRaw reads raw temperature (16-bit).
func (d *Device) ReadTemperatureRaw() (int16, error) {
	data, err := d.transport.ReadRegister(RegTEMP2, 2)
	if err != nil {
		return 0, err
	}
	return int16(data[0])<<8 | int16(data[1]), nil
}

// ReadTemperatureC reads temperature in degrees Celsius.
// Preliminary: T(°C) = raw / 100.0 + 25.0
func (d *Device) ReadTemperatureC() (float32, error) {
	raw, err := d.ReadTemperatureRaw()
	if err != nil {
		return 0, err
	}
	return float32(raw)/100.0 + 25.0, nil
}

// ReadStatus reads the status register.
func (d *Device) ReadStatus() (byte, error) {
	return d.readU8(RegSTATUS)
}

// ---------------------------------------------------------------------------
// Stateless conversion functions
// ---------------------------------------------------------------------------

// DecodeRaw20 decodes three bytes into a 20-bit two's complement integer.
func DecodeRaw20(b0, b1, b2 byte) int32 {
	raw := int32(b0)<<12 | int32(b1)<<4 | int32(b2)>>4
	if raw&0x80000 != 0 {
		raw -= 0x100000
	}
	return raw
}

// RawToG converts a decoded raw value to g.
func RawToG(raw int32, r Range) float32 {
	return float32(raw) * rangeToScale(r)
}

// RawToMps2 converts a decoded raw value to m/s².
func RawToMps2(raw int32, r Range) float32 {
	return float32(raw) * rangeToScale(r) * StandardGravityMS2
}

func rangeToScale(r Range) float32 {
	switch r {
	case Range2G:
		return Scale2GGPerLSB
	case Range4G:
		return Scale4GGPerLSB
	case Range8G:
		return Scale8GGPerLSB
	default:
		return Scale4GGPerLSB
	}
}

File: python/src/adxl355/device.py
"""Main ADXL355 device driver."""

from adxl355.constants import (
    DEVID_AD,
    DEVID_MST,
    PARTID,
    RESET_CODE,
    STANDARD_GRAVITY_M_S2,
    SCALE_2G_G_PER_LSB,
    SCALE_4G_G_PER_LSB,
    SCALE_8G_G_PER_LSB,
)
from adxl355.errors import (
    BusError,
    DeviceNotFoundError,
    InvalidConfigurationError,
)
from adxl355.registers import (
    Register,
    Range,
    PowerMode,
    ODR,
    RANGE_SEL_MASK,
    FILTER_ODR_MASK,
    FILTER_ODR_SHIFT,
    FILTER_HPF_MASK,
)
from adxl355.transport import Transport
from adxl355.types import RawXYZ, AccelXYZ


class ADXL355:
    """
    ADXL355 accelerometer driver.

    Transport-agnostic: accepts any object conforming to the Transport protocol.
    """

    def __init__(self, transport: Transport) -> None:
        self._transport = transport
        self._range = Range.G4
        self._initialized = False

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _read_reg(self, reg: int) -> int:
        data = self._transport.read_register(reg, 1)
        return data[0]

    def _write_reg(self, reg: int, value: int) -> None:
        self._transport.write_register(reg, bytes([value]))

    def _check_init(self) -> None:
        if not self._initialized:
            raise DeviceNotFoundError("Device not initialized. Call probe() first.")

    # ------------------------------------------------------------------
    # Core API
    # ------------------------------------------------------------------

    def probe(self) -> bool:
        """
        Verify device identity by reading ID registers.

        Returns True if all three ID registers match expected values.
        After successful probe, the device is left in standby mode.
        """
        id_ad = self._read_reg(Register.DEVID_AD)
        id_mst = self._read_reg(Register.DEVID_MST)
        part_id = self._read_reg(Register.PARTID)

        if id_ad != DEVID_AD or id_mst != DEVID_MST or part_id != PARTID:
            raise DeviceNotFoundError(
                f"Device ID mismatch: DEVID_AD=0x{id_ad:02X}, "
                f"DEVID_MST=0x{id_mst:02X}, PARTID=0x{part_id:02X}"
            )

        # Enter standby mode after probe
        self._write_reg(Register.POWER_CTL, PowerMode.STANDBY)
        self._initialized = True
        return True

    def reset(self) -> None:
        """Perform a software reset."""
        self._write_reg(Register.RESET, RESET_CODE)
        self._transport.delay_ms(10)
        self._range = Range.G4

    def set_range(self, range_val: Range) -> None:
        """Set the acceleration range.

        Datasheet Rev.D, Table 42: range in bits 1:0 (0x01=2g, 0x02=4g, 0x03=8g).
        Unrelated bits (INT_POL, I2C_HS) are preserved.
        """
        if range_val not in (Range.G2, Range.G4, Range.G8):
            raise InvalidConfigurationError(f"Invalid range: {range_val}")
        reg = self._read_reg(Register.RANGE)
        reg = (reg & ~RANGE_SEL_MASK) | (int(range_val) & RANGE_SEL_MASK)
        self._write_reg(Register.RANGE, reg)
        self._range = range_val

    def get_range(self) -> Range:
        """Read the currently configured range from hardware."""
        reg = self._read_reg(Register.RANGE)
        return Range(reg & RANGE_SEL_MASK)

    def set_power_mode(self, mode: PowerMode) -> None:
        """Set power mode (standby or measurement).

        Datasheet Rev.D, Table 43: bit 0 = 1 => standby, bit 0 = 0 => measurement.
        """
        reg = self._read_reg(Register.POWER_CTL)
        if mode == PowerMode.STANDBY:
            reg |= 1
        else:
            reg &= ~1
        self._write_reg(Register.POWER_CTL, reg)

    def set_odr(self, odr: ODR) -> None:
        """Set output data rate.

        Datasheet Rev.D, Table 38: ODR_LPF in bits 3:0, HPF_CORNER in bits 6:4.
        """
        if odr not in ODR.__members__.values():
            raise InvalidConfigurationError(f"Invalid ODR: {odr}")
        reg = self._read_reg(Register.FILTER)
        reg = (reg & FILTER_HPF_MASK) | ((int(odr) << FILTER_ODR_SHIFT) & FILTER_ODR_MASK)
        self._write_reg(Register.FILTER, reg)

    # ------------------------------------------------------------------
    # Data readout
    # ------------------------------------------------------------------

    def read_raw(self) -> RawXYZ:
        """Read raw 20-bit acceleration data for all three axes."""
        data = self._transport.read_register(Register.XDATA3, 9)
        x = _decode_raw20(data[0], data[1], data[2])
        y = _decode_raw20(data[3], data[4], data[5])
        z = _decode_raw20(data[6], data[7], data[8])
        return RawXYZ(x, y, z)

    def read_acceleration_g(self) -> AccelXYZ:
        """Read acceleration in g (gravity multiples)."""
        raw = self.read_raw()
        scale = _range_to_scale(self._range)
        return AccelXYZ(
            x=raw.x * scale,
            y=raw.y * scale,
            z=raw.z * scale,
        )

    def read_acceleration_mps2(self) -> AccelXYZ:
        """Read acceleration in m/s²."""
        accel = self.read_acceleration_g()
        return AccelXYZ(
            x=accel.x * STANDARD_GRAVITY_M_S2,
            y=accel.y * STANDARD_GRAVITY_M_S2,
            z=accel.z * STANDARD_GRAVITY_M_S2,
        )

    def read_temperature_raw(self) -> int:
        """Read raw temperature value (16-bit)."""
        data = self._transport.read_register(Register.TEMP2, 2)
        return (data[0] << 8) | data[1]

    def read_temperature_c(self) -> float:
        """
        Read temperature in degrees Celsius.

        Datasheet Rev.D: 12-bit unsigned, nominal intercept 1885 LSB at 25°C,
        slope -9.05 LSB/°C. Formula: T(°C) = 25.0 + (raw - 1885.0) / -9.05
        """
        raw = self.read_temperature_raw()
        return 25.0 + (raw - 1885.0) / -9.05

    def read_status(self) -> int:
        """Read the status register."""
        return self._read_reg(Register.STATUS)

    def read_fifo_entries(self) -> int:
        """Read the number of valid samples in the FIFO."""
        return self._read_reg(Register.FIFO_ENTRIES)


# ------------------------------------------------------------------
# Stateless conversion functions
# ------------------------------------------------------------------


def _decode_raw20(b0: int, b1: int, b2: int) -> int:
    """
    Decode three bytes into a 20-bit two's complement integer.

    Args:
        b0: MSB (first byte from XDATA3/YDATA3/ZDATA3)
        b1: Middle byte
        b2: LSB (last byte)

    Returns:
        Sign-extended integer in range [-524288, 524287]
    """
    raw = (b0 << 12) | (b1 << 4) | (b2 >> 4)
    if raw & 0x80000:
        raw -= 0x100000
    return raw


def raw_to_g(raw: int, range_val: Range) -> float:
    """Convert a decoded raw value to g."""
    return raw * _range_to_scale(range_val)


def raw_to_mps2(raw: int, range_val: Range) -> float:
    """Convert a decoded raw value to m/s²."""
    return raw * _range_to_scale(range_val) * STANDARD_GRAVITY_M_S2


def _range_to_scale(range_val: Range) -> float:
    if range_val == Range.G2:
        return SCALE_2G_G_PER_LSB
    elif range_val == Range.G4:
        return SCALE_4G_G_PER_LSB
    elif range_val == Range.G8:
        return SCALE_8G_G_PER_LSB
    return SCALE_4G_G_PER_LSB
